sklearn/tree/_tree.pyx

```diff
@@ -1132,3 +1132,284 @@ cdef class Tree:
+
+    # computes number of leaves in all branches and the overall impurity of
+    # the branch. The overall impurity is the sum of r_node in its leaves.
+    cdef:
+        SIZE_t[:] n_leaves = np.zeros(shape=n_nodes, dtype=np.intp)
+        DOUBLE_t[:] r_branch = np.zeros(shape=n_nodes, dtype=np.float64)
+        DOUBLE_t current_r
+        SIZE_t leaf_idx
+        SIZE_t parent_idx
+
+    with nogil:
+        for leaf_idx in range(leaves_in_subtree.shape[0]):
+            if leaves_in_subtree[leaf_idx] == 0:
+                continue
+            r_branch[leaf_idx] = r_node[leaf_idx]
+
+            # bubble up values to ancestor nodes
+            current_r = r_node[leaf_idx]
+            while leaf_idx != 0:
+                parent_idx = parent[leaf_idx]
+                r_branch[parent_idx] += current_r
+                n_leaves[parent_idx] += 1
+                leaf_idx = parent_idx
+
+    cdef:
+        # candidate nodes that can be pruned
+        bint[:] candidate_nodes = view.array(
+            shape=(n_nodes, ), itemsize=sizeof(bint), format="i")
+        # nodes in subtree
+        bint[:] in_subtree = view.array(
+            shape=(n_nodes, ), itemsize=sizeof(bint), format="i")
+        DOUBLE_t[:] g_node = np.zeros(shape=n_nodes, dtype=np.float64)
+        SIZE_t cur_idx
+        DOUBLE_t subtree_alpha
+        DOUBLE_t cur_ccp_alpha
+        SIZE_t child_l_idx
+        SIZE_t child_r_idx
+        SIZE_t n_pruned_leaves
+        DOUBLE_t r_diff
+        DOUBLE_t max_float64 = np.finfo(np.float64).max
+        SIZE_t init_capacity = 0
+
+    with nogil:
+        for i in range(leaves_in_subtree.shape[0]):
+            candidate_nodes[i] = not leaves_in_subtree[i]
+
+        for i in range(n_nodes):
+            in_subtree[i] = 1
+
+        while True:
+            # If root node is a leaf
+            if not candidate_nodes[0]:
+                break
+
+            # computes ccp_alpha for subtrees and finds the minimal alpha
+            cur_ccp_alpha = max_float64
+            for i in range(n_nodes):
+                if not candidate_nodes[i]:
+                    continue
+                subtree_alpha = (r_node[i] - r_branch[i]) / (n_leaves[i] - 1)
+                if subtree_alpha < cur_ccp_alpha:
+                    cur_ccp_alpha = subtree_alpha
+                    cur_idx = i
+
+            # The subtree on the previous iteration has the greatest ccp_alpha
+            # less than or equal to self.ccp_alpha
+            if cur_ccp_alpha > ccp_alpha:
+                break
+
+            # stack uses only the start variable
+            rc = stack.push(cur_idx, 0, 0, 0, 0, 0, 0)
+            if rc == -1:
+                with gil:
+                    raise MemoryError()
+
+            # descendants of branch are not in subtree
+            while not stack.is_empty():
+                stack.pop(&stack_record)
+                node_idx = stack_record.start
```

Comment:
As a slight optimization, would it be relevant to add:

```python
if not in_subtree[node_idx]:
    break  # branch has already been marked for pruning
```

?

The idea is that you might be trying to prune a branch that has already been pruned in at a previous iteration. In this case there's no need to go all the way down to the leaves of the original tree.